common/defs.go

```diff
@@ -82,6 +89,76 @@ type PortUpdateMessage struct {
 	Port uint64 `json:"port"`
 }
 
+type Endpoint struct {
+	CidrStr string `json:"cidr,omitempty"`
+	// TODO this can be collapsed into Cidr but needs
+	// work on JSON marshaller/unmarshaller to do that.
+	Cidr      net.IPNet `json:"-"`
+	TenantID  uint64    `json:"tenant_id,omitempty"`
+	TenantExternalID string  `json:"tenant_external_id,omitempty"`
+	SegmentID uint64    `json:"segment_id,omitempty"`
+	SegmentExternalID string `json:"segment_external_id,omitempty"`
+}
+
+const (
+	PolicyDirectionIngress = "ingress"
+	PolicyDirectionEgress  = "egress"
+)
+
+type PortRange [2]uint
+
+type Rule struct {
+	Protocol   string
+	Ports      []uint
+	PortRanges []PortRange
+	IcmpType   uint
+	IcmpCode   uint
+	Stateful   bool
+}
+
+type Rules []Rule
+
+// Metadata attached to entities for various external environments like openstack/kybernetes
+type Tag struct {
+	Key   string
+	Value string
+}
+
+type Policy struct {
+	Direction   string     `json:"direction,omitempty"`
+	Description string     `json:"description,omitempty"`
+	Name        string     `json:"name,omitempty"`
+	Id          string     `json:"id,omitempty"`
+	AppliedTo   []Endpoint `json:"applied_to,omitempty"`
+	Peers       []Endpoint `json:"peers,omitempty"`
+	Rules       []Rule     `json:"rule,omitempty"`
+	Tags        []Tag      `json:"tags,omitempty"`
+}
+
+func (p *Policy) Validate() error {
+	errMsg := ""
+	p.Direction = strings.ToLower(p.Direction)
+	if p.Direction != PolicyDirectionEgress && p.Direction != PolicyDirectionIngress {
+		errMsg += fmt.Sprintf("Unknown direction '%s', allowed '%s' or '%s'. ", p.Direction, PolicyDirectionEgress, PolicyDirectionIngress)
+	}
+	for _, r := range p.Rules {
+		r.Protocol = strings.ToLower(r.Protocol)
+		if r.Protocol == "tcp" {
+			r.Stateful = true
```

Comment:
Oh, this is just the beginning. It is by no means full validation method yet. But I'll make all changes mentioned here.